ddl/ddl_tiflash_api.go

```diff
@@ -57,22 +57,156 @@ type TiFlashReplicaStatus struct {
+type PollTiFlashBackoffContext struct {
+	MinTick TiFlashTick
+	MaxTick TiFlashTick
+	// Capacity limits tables a backoff pool can handle, in order to limit handling of big tables.
+	Capacity int
+	Rate     TiFlashTick
+	elements map[int64]*PollTiFlashBackoffElement
+}
+
+// NewPollTiFlashBackoffContext creates an instance of PollTiFlashBackoffContext.
+func NewPollTiFlashBackoffContext(MinTick, MaxTick TiFlashTick, Capacity int, Rate TiFlashTick) (*PollTiFlashBackoffContext, error) {
+	if MaxTick < MinTick {
+		return nil, fmt.Errorf("`MaxTick` should always be larger than `MinTick`")
+	}
+	if MinTick < 1 {
+		return nil, fmt.Errorf("`MinTick` should not be less than 1")
+	}
+	if Capacity < 0 {
+		return nil, fmt.Errorf("negative `Capacity`")
+	}
+	if Rate <= 1 {
+		return nil, fmt.Errorf("`Rate` should always be larger than 1")
+	}
+	return &PollTiFlashBackoffContext{
+		MinTick:  MinTick,
+		MaxTick:  MaxTick,
+		Capacity: Capacity,
+		elements: make(map[int64]*PollTiFlashBackoffElement),
+		Rate:     Rate,
+	}, nil
+}
+
 // TiFlashManagementContext is the context for TiFlash Replica Management
 type TiFlashManagementContext struct {
 	TiFlashStores             map[int64]helper.StoreStat
 	HandlePdCounter           uint64
 	UpdateTiFlashStoreCounter uint64
 	UpdateMap                 map[int64]bool
+	Backoff                   *PollTiFlashBackoffContext
+}
+
+// Tick will first check increase Counter.
+func (b *PollTiFlashBackoffContext) Tick(ID int64) (bool, bool, int) {
+	e, ok := b.Get(ID)
+	if !ok {
+		return false, false, 0
+	}
+	growed := e.MaybeGrow(b)
+	e.Counter += 1
+	e.TotalCounter += 1
+	return growed, true, e.TotalCounter
+}
+
+func (e *PollTiFlashBackoffElement) limit() int {
+	return int(e.Threshold)
+}
+
+// NeedGrow returns if we need to grow
+func (e *PollTiFlashBackoffElement) NeedGrow() bool {
+	return e.Counter >= e.limit()
+}
+
+func (e *PollTiFlashBackoffElement) doGrow(b *PollTiFlashBackoffContext) {
+	defer func() {
+		e.Counter = 0
+	}()
+	if e.Threshold < b.MinTick {
+		e.Threshold = b.MinTick
+	}
+	if e.Threshold*b.Rate > b.MaxTick {
+		e.Threshold = b.MaxTick
+	} else {
+		e.Threshold *= b.Rate
+	}
+}
+
+// MaybeGrow grows threshold and reset counter when need
+func (e *PollTiFlashBackoffElement) MaybeGrow(b *PollTiFlashBackoffContext) bool {
+	if !e.NeedGrow() {
+		return false
+	}
+	e.doGrow(b)
+	return true
+}
+
+// Remove will reset table from backoff
+func (b *PollTiFlashBackoffContext) Remove(ID int64) bool {
+	_, ok := b.elements[ID]
+	delete(b.elements, ID)
+	return ok
+}
+
+// Get returns pointer to inner PollTiFlashBackoffElement.
+// Only exported for test.
+func (b *PollTiFlashBackoffContext) Get(ID int64) (*PollTiFlashBackoffElement, bool) {
+	res, ok := b.elements[ID]
+	return res, ok
+}
+
+// Put will record table into backoff pool, if there is enough room, or returns false.
+func (b *PollTiFlashBackoffContext) Put(ID int64) (*PollTiFlashBackoffElement, bool) {
```

Comment:
Why do we need to return `*PollTiFlashBackoffElement` here?

Reply:
Seem to be useless currently, gonna remove.